doc/sphinx/src/using.rst

```diff
@@ -73,7 +73,42 @@ limited. The syntax is:
 where ``Function`` now takes as many indices are required and
 ``reduced`` as arguments.
 
-portable_arrays.hpp
+portable_errors.hpp
+^^^^^^^^^^^^^^^^^^^^
+
+``portable_errors.hpp`` provides error handling that works with
+different portability backends, such as `Kokkos`. We provide several
+useful macros. All the macros in this file will print the file and
+line number where the macro was called, enabling easier debugging.
+
+The following macros are **disabled** automaticaly for production
+builds (e.g., when the ``NDEBUG`` preprocessor macro is defined):
+
+* ``PORTABLE_REQUIRE(condition, message)`` prints an error message and aborts the program (without throwing an exception) if compiled in debug mode and ``condition`` is not satisfied. This macro **works** in device kernels.
+* ``PORTABLE_REQUIRE_THROWS(condition, message)`` raises a runtime exception (annotated with the message) when compiled in debug mode and the condition is not satisfied. This macro **does not** work in device kernels.
```

Comment:
Given that many of these _do not_ work on device, what then is the definition of "portable"?

Reply:
In this case I'm just namespacing, because the macros can't live in a `PortsOfCall` namespace, so really `PORTABLE` here is `PORTS_OF_CALL`

I think it makes sense to provide a unified interface though with the truly portable versions and the not-so-portable versions that use exception handling. The issue is exception handling simply does not work on device, but we may still want to provide it. Here by exception handling I mean specifically the machinery provided by `stdexcept` which provides things like
```C++
try {
  // some code
  if (condition) {
    PORTABLE_THROW(message)
  } catch (const std::runtime_error &error) {
    // some code that recovers from the error gracefully
  }
}
```
what `ports-of-call` is providing on top of that is a filename and line number, as well as `THROW_OR_ABORT` which gracefully switches between the exception model and the "print an error and die" model.

Reply:
I suppose we could not provide `REQUIRE_THROWS` or `THROWS` and only provide `THROW_OR_ABORT`. But I think sometimes we actually do want to just throw.

Reply:
> This will be very helpful I think. But rather than some of these just not working on device, should they instead do nothing? My issue is that if a "portable" implementation doesn't work on device, it begs the question as to whether it's truly portable anymore.

I recognize that having a *universal* error handling mechanism is probably a pipe dream, but is it not possible to have something that could yet function on device and do something?  I seem to think that we've had things that can abort a kernel and then set some error state (which may not be seen until the next kernel launch) but might be limited to CUDA or something like that (which, to be clear, we should support).  I'm certainly no expert on error handling!  I think we should be careful about putting together something that could easily be snuck in to some inappropriate parts of library codes that would then make for some downstream headaches; at the very least if the macros have to expand to nothing on device that would be better than having them expand to something with undefined behavior.

Reply:
It is possible to have something that throws an error on host and aborts on device. I could remove the functions that only work on host if that's how people feel, and leave things that are all truly portable. The only disadvantage of this is that we will be excluding any use of C++ exceptions if anyone cares about that. But I suppose if someone wants to use that it can be on them.

Reply:
My main concern is the case where a downstream library puts in this kind of functionality to ensure certain behavior, but in reality this is device code (at least sometimes) and the behavior becomes either undefined or wrong/unexpected.  I can imagine it being a very hard bug to track down if there is some kind of silent-ish abort of a device-side constructor or something similar but the host side is never the wiser.  

Maybe I'm being a little too paranoid about this here?

Reply:
so I *think* the functionality that doesn't work on device will fail to compile. But I should double check that. We *can* simply not provide the functionality that might fail. The reason I'm reluctant to do that is I feel that there are times when C++ exceptions provide a useful ability on host.

The aborts that *do* run on device will all run on host or device and will be loud.

Reply:
> In this case I'm just namespacing, because the macros can't live in a `PortsOfCall` namespace, so really `PORTABLE` here is `PORTS_OF_CALL`
> 
> I think it makes sense to provide a unified interface though with the truly portable versions and the not-so-portable versions that use exception handling. The issue is exception handling simply does not work on device, but we may still want to provide it. Here by exception handling I mean specifically the machinery provided by `stdexcept` which provides things like
> 
> ```c++
> try {
>   // some code
>   if (condition) {
>     PORTABLE_THROW(message)
>   } catch (const std::runtime_error &error) {
>     // some code that recovers from the error gracefully
>   }
> }
> ```
> 
> what `ports-of-call` is providing on top of that is a filename and line number, as well as `THROW_OR_ABORT` which gracefully switches between the exception model and the "print an error and die" model.

But what is the true _utililty_ of `PORTABLE_THROW`? What if I modify the above code (I assume it should be try/catch rather than try{if/catch}, but my C++ knowledge is limited) to be
```c++
try {
   // some code
   if (condition) {
      const char* const message = PORTABLE_ERR_MSG(message)
      std::runtime_error(msg)
   }
} catch (const std::runtime_error &error) {
   // some code that recovers from the error gracefully
}
```
If the utility is in the creation of the error message, should that then be the functionality that is exposed?

Reply:
Hmm. Yeah that's a good point. I like the idea of exposing the error message generation and removing the throw machinery... with the exception of the macro that either throws or aborts depending.

Reply:
Ok I've made the change. Here's how it works now:
- `PORTABLE_THROW` is gone.
- `PORTABLE_THROW_OR_ABORT` still exists and is the preferred way to perform checks that you would like to throw an exception but want to work on device.
- I have added `PORTABLE_ERROR_MESSAGE` . It can't work as cleanly as your suggested code above @jhp-lanl because that kind of assignment leaks memory, so you have to fill an array that is passed in as an argument. It also has to be a `char*` to work on device, which means the user is responsible for allocating enough memory to support their string on device without a segfault. But I think that's fine.

Reply:
I was thinking about this yesterday but forgot to chime in. Would a compile time error message length solve the memory leak issue? Could the length default to something and be set via CMake?

BUUUUUT, I don't feel strongly. Simply printing something and then raising an exception is also fine with me.

Reply:
> I was thinking about this yesterday but forgot to chime in. Would a compile time error message length solve the memory leak issue? Could the length default to something and be set via CMake?

That would allow us to fix length ahead of time, which would avoid the segfault thing, but it wouldn't allow us to return a pointer. We'd still have to fill an array that's passed to us. I think the current design is fine, though I'm not opposed to a compile-time string length if people feel strongly.